### ISSUE TITLE
borg restore of hard-linked file in proxmox lxc container fails under some circumstances

### ISSUE THREAD

**Issue by oddfellow** [borgbackup/borg#4535]
## Have you checked borgbackup docs, FAQ, and open Github issues?

yes

## Is this a BUG / ISSUE report or a QUESTION?

BUG / ISSUE

#### Your borg version (borg -V).

borg 1.1.9 (standalone binary, both client and server)

#### Operating system (distribution) and version.

Client: Ubuntu 18.04.2 LTS (Bionic Beaver) amd64
Server: Debian GNU/Linux 9 (stretch) amd64

#### Hardware / network configuration, and filesystems used.

Client: Proxmox 5.3 LXC container (unprivileged)
Server: Debian 9 amd64 bare-metal

#### How much data is handled by borg?

for the test case only a text file

#### Full borg commandline that lead to the problem (leave away excludes and passwords)

borg extract ::test1 root/baktest/tmp2

## Describe the problem you're observing.

On a Proxmox 5 LXC container I do a backup of a directory with subdirectories with some hard-linked files (the real setup is a cyrus mail store). If I restore only one sub-directory with a hard-linked file to a file in another dir which is not restored, the borg process on the client hangs forever with 100% cpu usage. The restored file is created with 0 bytes.

This happens only in the LXC container. If I do the restore on a non-LXC system, e.g. on the backup server, the restore works as expected.

A restore of the whole archive also works without problems and the hard-links are re-created.

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

The problem can be reproduced with:

- create a proxmox lxc container with Ubuntu 18.04 template
- copy borg standalone binary to container
- ssh into container
- mkdir /root/baktest/{tmp1,tmp2}
- echo "test" > /root/baktest/tmp1/hl1.txt
- ln /root/baktest/tmp1/hl1.txt /root/baktest/tmp2/hl2.txt
- borg init --encryption=repokey ssh://...
- borg create ::test1 /root/baktest/
- borg extract ::test1 /root/baktest/tmp2

#### Include any warning/errors/backtraces from the system logs

no warnings, no errors visible

**Comment by ThomasWaldmann** [comment]
Interesting.

Can you extract tmp1 in the container?

Can you do a `strace -f` of the extract run that hangs forever?

**Comment by oddfellow** [comment]
Thanks for your reply.
Yes I can extract tmp1.
The strace file is attached.
[borg-strace.txt](https://github.com/borgbackup/borg/files/3131027/borg-strace.txt)

**Comment by ThomasWaldmann** [comment]
```
40224 lstat("/root/root/baktest/tmp2", 0x7fff59f92720) = -1 ENOENT (No such file or directory)
40224 stat("/root/root/baktest", 0x7fff59f922a0) = -1 ENOENT (No such file or directory)
40224 stat("/root/root", 0x7fff59f92010) = -1 ENOENT (No such file or directory)
40224 stat("/root", {st_mode=S_IFDIR|0700, st_size=4096, ...}) = 0
40224 mkdir("/root/root", 0777)         = 0
40224 mkdir("/root/root/baktest", 0777) = 0
40224 stat("/root/root/baktest/tmp2", 0x7fff59f92530) = -1 ENOENT (No such file or directory)
40224 mkdir("/root/root/baktest/tmp2", 0777) = 0
40224 lstat("/root/root/baktest/tmp2/hl2.txt", 0x7fff59f92720) = -1 ENOENT (No such file or directory)
40224 stat("/root/root/baktest/tmp2", {st_mode=S_IFDIR|0700, st_size=4096, ...}) = 0
40224 openat(AT_FDCWD, "/root/root/baktest/tmp2/hl2.txt", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = 4
40224 fstat(4, {st_mode=S_IFREG|0600, st_size=0, ...}) = 0
40224 ioctl(4, TCGETS, 0x7fff59f924b0)  = -1 ENOTTY (Inappropriate ioctl for device)
40224 lseek(4, 0, SEEK_CUR)             = 0
40224 select(9, [6 8], [5], [5 6 8], {tv_sec=1, tv_usec=0}) = 1 (out [5], left {tv_sec=0, tv_usec=999998})
# the strace ends here, so this is the "hang"
```

**Comment by oddfellow** [comment]
Yes. Do you need the rest of the strace after pressing ctrl-c? Or something else?

**Comment by ThomasWaldmann** [comment]
Yes, a complete strace would be good.

**Comment by oddfellow** [comment]
[borg-strace.txt](https://github.com/borgbackup/borg/files/3131391/borg-strace.txt)

**Comment by ThomasWaldmann** [comment]
OK, so after the stuff in the strace above, we have just this:
```
43216 <... wait4 resumed> 0x7ffdc4ceb3ec, 0, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
43220 <... select resumed> )            = ? ERESTARTNOHAND (To be restarted if no handler)
43216 --- SIGINT {si_signo=SIGINT, si_code=SI_KERNEL} ---
43220 --- SIGINT {si_signo=SIGINT, si_code=SI_KERNEL} ---
43216 kill(43217, SIGINT <unfinished ...>
```

So it really hangs in the `select`.

Hanging while extracting hardlinks reminds me of #4350 - this might be a duplicate of that.

**Comment by ThomasWaldmann** [comment]
@oddfellow thanks for this bugreport, the detail that it only happens for partial extraction really helped to find what the root cause is. i am quite confident that this is the same issue as seen in #4350 and fixed by #4539, so i am closing this issue now as a duplicate.

i didn't yet merge this into master branch or backport it yet, not sure if the changeset applies cleanly to 1.1-maint so you could try it.
